### PR TITLE
Support for commands in parallel

### PR DESCRIPTION
**Pull request by julianoes**
This is a big refactor of the MavlinkCommands class in order to support
multiple commands in parallel. It won't do multiple commands with the
same command ID at the same time in order to prevent conflicts/races
with acks that can't be assigned uniquely.

### CODE REVIEW

File: .ycm_extra_conf.py [hunk 1]
```diff
@@ -0,0 +1,206 @@
+# This file is NOT licensed under the GPLv3, which is the license for the rest
+# of YouCompleteMe.
```
Comment (hamishwillee) [inline .ycm_extra_conf.py]: See YouCompleteMe

Comment (julianoes) [reply .ycm_extra_conf.py]: 😄 what?

Comment (hamishwillee) [reply .ycm_extra_conf.py]: Exactly - what!!

The text says "This file is NOT licensed under the GPLv3, which is the license for the rest of YouCompleteMe."

That looks like a placeholder to me.

And it should be obvious I am not actually reviewing this - I just wanted to see what the code looked like :-)

Comment (julianoes) [reply .ycm_extra_conf.py]: I can remove it and say it is licensed as BSD 3-clause like the rest of the repo.

Comment (hamishwillee) [reply .ycm_extra_conf.py]: Thanks.

Comment (julianoes) [reply .ycm_extra_conf.py]: Actually this file says it is public domain, so I'm not sure what I should do.

Comment (hamishwillee) [reply .ycm_extra_conf.py]: I thought this was our file. If YouComleteMe is an actual real name of a product called YouCompleteMe then we have to leave that alone - it isn't ours to modify.

File: core/log.h [hunk 2]
```diff
@@ -101,6 +102,7 @@ class LogDetailed
                 std::cout << "|Debug] ";
                 break;
             case LogLevel::Info:
+                std::cout << ANSI_COLOR_BLUE;
```
Comment (darioxz) [inline core/log.h]: This is really not critical, but #86 should get merged before #86.

File: .ycm_extra_conf.py (hunk 1 above)
Comment (mrpollo) [reply .ycm_extra_conf.py]: Yep this file was copied from the YouCompleteMe project, we should leave as is

File: core/log.h (hunk 2 above)
Comment (mrpollo) [reply core/log.h]: Is #86 rebased on to of #85?, can we remove this so we can merge w/o blocking?

File: core/locked_list.h [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+public:
+    LockedList() :
+        _list(),
+        _mutex()
+    {};
+
+    class iterator: public std::list<T>::iterator
+    {
+    public:
+        iterator(typename std::list<T>::iterator c, std::mutex &mutex) :
```
Comment (mrpollo) [inline core/locked_list.h]: I know we are using single char variables elsewhere but we should move away from this, can we name this variable something more descriptive, please?

File: core/mavlink_commands.h [hunk 4]
```diff
@@ -53,30 +54,27 @@ class MavlinkCommands
...
 private:
-    enum class State {
-        NONE,
-        WAITING,
-        IN_PROGRESS,
-        DONE,
-        FAILED
-    } _state = State::NONE;
-    std::mutex _state_mutex {};
+    static constexpr double DEFAULT_TIMEOUT_NORMAL_S = 0.5;
```
Comment (mrpollo) [inline core/mavlink_commands.h]: We need to find a way to get the default timeouts into the documentation (if they aren't already there), this is important behavior that should be visible to external developers

File: core/mavlink_commands.h [hunk 5]
```diff
@@ -53,30 +54,27 @@ class MavlinkCommands
...
+    static constexpr double DEFAULT_TIMEOUT_IN_PROGRESS_S = 3.0;
+    static constexpr int RETRIES = 3;
 
     struct Work {
-        int retries_to_do = 3;
-        double timeout_s = 0.5;
+        int num_sent = 0;
```
Comment (mrpollo) [inline core/mavlink_commands.h]: can we use a more descriptive name here? is `num_set` the number of times a command has been set after retries? I tried following the code and got a bit confused

File: core/timeout_handler.cpp [hunk 6]
```diff
@@ -35,8 +35,17 @@ void TimeoutHandler::refresh(const void *cookie)
 
     auto it = _timeouts.find((void *)(cookie));
     if (it != _timeouts.end()) {
-        dl_time_t future_time = steady_time_in_future(it->second->duration_s);
-        it->second->time = future_time;
+        it->second->start_time = steady_time();
+    }
+}
+
+void TimeoutHandler::update(double new_duration_s, const void *cookie)
```
Comment (mrpollo) [inline core/timeout_handler.cpp]: what about `updated_duration_s` instead of `new_duration_s`?

File: core/mavlink_commands.h (hunk 4 above)
Comment (hamishwillee) [reply core/mavlink_commands.h]: @julianoes @mrpollo Arguably some timeouts should actually be part of the settable public API - and if so they could then be documented in the same way as everything else (?).

@julianoes Are all timeouts named consistently so we can find them? (i.e. can I have a list)

File: core/locked_list.h (hunk 3 above)
Comment (julianoes) [reply core/locked_list.h]: Ok 173ac1a.

File: core/mavlink_commands.h (hunk 4 above)
Comment (julianoes) [reply core/mavlink_commands.h]: Ok, I agree. Let's check this later.

File: core/timeout_handler.cpp (hunk 6 above)
Comment (julianoes) [reply core/timeout_handler.cpp]: Sure, thanks. See ea4cdb8.

File: core/mavlink_commands.h (hunk 5 above)
Comment (julianoes) [reply core/mavlink_commands.h]: It is `num_sent`, not `num_set` but I've changed it to `num_command_sent` in a746eab.

**Pull request merged**